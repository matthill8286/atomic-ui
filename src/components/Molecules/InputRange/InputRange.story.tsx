import { withState } from '@dump247/storybook-state'
import { boolean, number, select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import React from 'react'
import { InputRange, InputRangeProps } from './InputRange'

const stories = storiesOf('Design System/Molecules/InputRange/Basic', module)

stories.add(
  'default',
  withState({ value: '0.5' }, store => {
    const onChange = event => {
      const {
        target: { value },
      } = event
      store.set({ value })
    }

    const knobs = (): InputRangeProps => ({
      min: text('min', '0'),
      max: text('max', '1'),
      step: text('step', '0.1'),
      markAmount: number('markAmount', 1),
      showMarks: boolean('Show Marks', true),
      withBubble: boolean('With Bubble', true),
      allowMoreThanMax: boolean('allowMoreThanMax', false),
    })

    return <InputRange value={store.state.value} onChange={onChange} {...knobs()} />
  })
)
