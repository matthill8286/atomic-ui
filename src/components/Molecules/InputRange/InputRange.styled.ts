import { css, media, saiyanTheme, styled } from '@/styles'
import { Typo } from '../../Atoms/Typography'

const issaiyan = theme => theme.name === saiyanTheme.name

export const StyledInputRangeContainer = styled.div`
  padding: ${({ theme }) => theme.spacing.base.md};
  flex-grow: 1;
  ${media.maxSm} {
    padding: ${({ theme }) => `${theme.spacing.base.md} ${theme.spacing.base.lg}`};
  }
`
export const StyledInputRangeBubbleContainer = styled.div`
  height: 6rem;
`

export const Animation = ({ theme }) => css`
  animation: wobble 2s ease-in-out 1 alternate;

  @keyframes wobble {
    0% {
      transform: scale(1);
    }
    20% {
      transform: scale(1.15);
    }
    30% {
      transform: scale(1);
    }
    40% {
      transform: scale(1.1);
    }
    50% {
      transform: scale(1);
    }
    60% {
      transform: scale(1.05);
    }
    70% {
      transform: scale(1);
    }
    80% {
      transform: scale(1.02);
    }
    100% {
      transform: scale(1);
    }
  }
`

export const StyledInputRangeBubblePosition = styled.output`
  ${({ theme }) =>
    css`
      width: ${issaiyan(theme) ? `4rem` : `3.5rem`};
      height: ${issaiyan(theme) ? `4rem` : `3.5rem`};

      position: relative;
      display: flex;
      justify-content: center;
      align-items: center;
      transform: translateX(-50%) ${!issaiyan(theme) && ` rotate(45deg)`};

      ${media.maxMd} {
        transform: translateX(-50%)
          ${issaiyan(theme)
            ? ` translateY(${theme.spacing.base.sm})`
            : ` translateY(${theme.spacing.base.xs}) rotate(45deg)`};
      }
    `}
`

export const StyledInputRangeBubble = styled.output<{ isSliding: boolean }>`
  ${({ theme, isSliding }) =>
    css`
      box-sizing: border-box;
      width: 100%;
      height: 100%;
      display: flex;
      justify-content: center;
      align-items: center;

      border: 2px solid ${theme.color.primary};
      ${issaiyan(theme) && `border-radius: 50%;`}
      ${isSliding && Animation}
    `}
`

export const StyledInputRangeBubbleTypo = styled(Typo)`
  ${({ theme }) =>
    css`
      font-size: ${!issaiyan(theme) ? theme.font.size.md : theme.font.size.xl};
      transform: ${!issaiyan(theme) ? ` translateY(4px) rotate(-45deg)` : `translateY(0)`};
      line-height: ${theme.font.lineHeight.xl};
      padding: 0;

      ${media.maxMd} {
        font-size: ${!issaiyan(theme) ? theme.font.size.xs : theme.font.size.md};
      }
    `}
`

export const StyledInputRangeThumb = styled.div`
  -webkit-appearance: none;
  box-shadow: ${({ theme }) => theme.dimension.elevationLevel1};
  height: 1px;
  width: 1px;
  background: ${({ theme }) => theme.color.primary};
  cursor: pointer;
  position: relative;
  &:after {
    content: '';
    position: absolute;
    height: ${({ theme }) => theme.spacing.base.md};
    width: ${({ theme }) => theme.spacing.base.md};
    margin-left: -${({ theme }) => theme.spacing.base.sm};
    border-radius: 50%;
    background: ${({ theme }) => theme.color.white};
    border: ${({ theme }) => `5px solid ${theme.color.primary}`};
  }
`

const InputRangeThumbArea = ({ theme }) => css`
  /* WebKit/Blink */
  ::-webkit-slider-thumb {
    -webkit-appearance: none;
    height: 1px;
    width: 1px;
    border-radius: 50%;
    border: 2px solid ${theme.color.white};
    transform: translateY(${theme.spacing.base.xxs}) scale(${theme.defaultSpacing * 4});
    ${media.maxMd} {
      transform: translateY(${theme.spacing.base.xxs}) scale(${theme.defaultSpacing * 2});
    }
    opacity: 0;
    cursor: pointer;
  }

  /* Firefox */
  ::-moz-range-thumb {
    -webkit-appearance: none;
    height: 1px;
    width: 1px;
    border: 2px solid ${theme.color.white};
    transform: scale(${theme.defaultSpacing * 4});
    border-radius: 50%;
    opacity: 0;
    border: none;
    cursor: pointer;
  }

  /* IE */
  ::-ms-thumb {
    -webkit-appearance: none;
    height: 1px;
    width: 1px;
    border: 2px solid ${theme.color.white};
    border-radius: 50%;
    opacity: 0;
    cursor: pointer;
  }
`

const StyledInputRangeTrack = ({ theme }) => css`
  ::-webkit-slider-runnable-track {
    width: 100%;
    height: ${theme.spacing.base.xs};
    cursor: pointer;
    background: ${theme.color.grey2};
    box-shadow: ${theme.dimension.elevationLevel1};
    border-radius: ${theme.dimension.borderRadius1};
  }

  :focus::-webkit-slider-runnable-track {
    background: ${theme.color.grey2};
  }

  ::-moz-range-track {
    width: 100%;
    height: ${theme.spacing.base.xs};
    cursor: pointer;
    background: ${theme.color.grey2};
    box-shadow: ${theme.dimension.elevationLevel1};
    border-radius: ${theme.dimension.borderRadius1};
  }

  ::-ms-track {
    width: 100%;
    height: ${theme.spacing.base.xs}px;
    cursor: pointer;
    background: transparent;
    border-color: transparent;
    border-width: 16px 0;
    color: transparent;
  }

  ::-ms-fill-lower {
    background: ${theme.color.grey2};
    box-shadow: ${theme.dimension.elevationLevel1};
    border-radius: ${theme.dimension.borderRadius1};
  }

  ::-ms-fill-upper {
    background: ${theme.color.grey2};
    box-shadow: ${theme.dimension.elevationLevel1};
    border-radius: ${theme.dimension.borderRadius1};
  }

  :focus::-ms-fill-lower {
    background: ${theme.color.grey2};
  }

  :focus::-ms-fill-upper {
    background: ${theme.color.grey2};
  }
`

export const StyledInputRange = styled.input`
  width: 100%;
  height: ${({ theme }) => theme.spacing.base.lg};
  -webkit-appearance: none;
  :focus {
    outline: none;
  }

  ${InputRangeThumbArea}
  ${StyledInputRangeTrack}
`

export const StyledInputRangeMarks = styled.div`
  position: relative;
  padding-bottom: 4rem;
`

export const StyledInputRangeMarkLine = styled.div<{
  left: number
  index: number
  markAmount: number
}>`
  left: ${({ left }) => left}%;
  top: -4px;
  position: absolute;
  height: ${({ theme }) => theme.spacing.base.md};
  width: 1px;
  background-color: ${({ theme }) => theme.color.grey3};

  ${media.maxMd} {
    display: ${({ index, markAmount }) => {
      const showFirstLastAndEverySecond = index === 0 || index === markAmount || index % 2 === 0
      return showFirstLastAndEverySecond ? 'block' : 'none'
    }};
  }

  ${media.maxSm} {
    display: ${({ index, markAmount }) => {
      const showFirstLastAndEveryFifth = index === 0 || index === markAmount || index % 5 === 0
      return showFirstLastAndEveryFifth ? 'block' : 'none'
    }};
  }
`

export const StyledInputRangeMarkTypo = styled(Typo)`
  position: absolute;
  max-width: unset;
  overflow: unset;
  text-overflow: ellipsis;
  top: ${({ theme }) => theme.spacing.base.md};
  transform: translateX(-50%);
`
