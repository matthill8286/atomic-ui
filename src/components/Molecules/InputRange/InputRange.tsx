import React, { useState } from 'react'
import {
  StyledInputRange,
  StyledInputRangeBubble,
  StyledInputRangeBubbleContainer,
  StyledInputRangeBubblePosition,
  StyledInputRangeBubbleTypo,
  StyledInputRangeContainer,
  StyledInputRangeMarkLine,
  StyledInputRangeMarks,
  StyledInputRangeMarkTypo,
  StyledInputRangeThumb,
} from './InputRange.styled'

export type InputRangeProps = Pick<
  JSX.IntrinsicElements['input'],
  'name' | 'step' | 'min' | 'max' | 'value'
> & {
  markAmount?: number
  allowMoreThanMax?: boolean
  withBubble?: boolean
  showMarks?: boolean
  onChange?: React.ChangeEventHandler<HTMLInputElement>
}

export const InputRange: React.FC<InputRangeProps> = props => {
  const {
    name = 'input-range',
    step = '0.1',
    markAmount = 0,
    min = '0',
    max = '1',
    value = '1',
    allowMoreThanMax = false,
    withBubble = true,
    showMarks = false,
    onChange,
  } = props

  const [isSliding, setIsSliding] = useState(false)
  const percentage = ((+value - +min) * 100) / (+max - +min)

  const onChangeHandler = (ev: React.ChangeEvent<HTMLInputElement>) => {
    if (onChange) {
      onChange(ev)
    }
  }

  const calcIntervalLabel = (interval: number): number =>
    Math.round(interval * ((+max - +min) / markAmount) + +min)

  return (
    <StyledInputRangeContainer>
      {withBubble && (
        <StyledInputRangeBubbleContainer>
          <StyledInputRangeBubblePosition style={{ left: `${percentage}%` }}>
            <StyledInputRangeBubble isSliding={isSliding}>
              <StyledInputRangeBubbleTypo
                color="primary"
                fontFamily="default"
                limitLines={1}
                tag="span"
                weight="regular">
                {Math.round(parseFloat(value as string) * 10)}
              </StyledInputRangeBubbleTypo>
            </StyledInputRangeBubble>
          </StyledInputRangeBubblePosition>
        </StyledInputRangeBubbleContainer>
      )}
      <StyledInputRangeThumb style={{ left: `${percentage}%` }} />
      <StyledInputRange
        type="range"
        name={name}
        list={name}
        step={step}
        min={min}
        max={max}
        value={value}
        onMouseDown={() => setIsSliding(true)}
        onTouchStart={() => setIsSliding(true)}
        onMouseUp={() => setIsSliding(false)}
        onTouchEnd={() => setIsSliding(false)}
        onChange={onChangeHandler}
      />
      {showMarks && markAmount > 0 && (
        <StyledInputRangeMarks>
          {Array.from({ length: markAmount + 1 }, (_, interval) => (
            <StyledInputRangeMarkLine
              key={interval}
              index={interval}
              markAmount={markAmount}
              left={((calcIntervalLabel(interval) - +min) * 100) / (+max - +min)}>
              <StyledInputRangeMarkTypo weight="semibold" fontSize="lg" limitLines={1}>
                {calcIntervalLabel(interval)} {allowMoreThanMax && interval === markAmount && '+'}
              </StyledInputRangeMarkTypo>
            </StyledInputRangeMarkLine>
          ))}
        </StyledInputRangeMarks>
      )}
    </StyledInputRangeContainer>
  )
}
