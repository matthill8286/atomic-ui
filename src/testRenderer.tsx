import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { Theme } from '@/types'

export function renderWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>,
  theme = saiyanTheme
) {
  return renderer.create(<ThemeProvider theme={theme}>{tree}</ThemeProvider>).toJSON()
}

export function renderWithThemeAndRouter(tree: React.ReactElement, theme: Theme = saiyanTheme) {
  return renderer
    .create(
      <ThemeProvider theme={theme}>
        <Router>{tree}</Router>
      </ThemeProvider>
    )
    .toJSON()
}

export function renderWithRouter(tree: React.ReactElement) {
  return renderer.create(<Router>{tree}</Router>).toJSON()
}

export function mountWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(<ThemeProvider theme={saiyanTheme}>{tree}</ThemeProvider>)
}

export function mountWithThemeAndRouter<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(
    <ThemeProvider theme={saiyanTheme}>
      <Router>{tree}</Router>
    </ThemeProvider>
  )
}
