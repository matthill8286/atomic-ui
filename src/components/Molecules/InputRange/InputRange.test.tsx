import React from 'react'
import { mountWithTheme, renderWithTheme } from '@/testRenderer'
import { InputRange } from './InputRange'

describe('InputRange', () => {
  it('should render', () => {
    const wrapper = renderWithTheme(<InputRange />)
    expect(wrapper).toMatchSnapshot()
  })
  it('should fire change event', () => {
    const onChange = jest.fn()
    const wrapper = mountWithTheme(<InputRange onChange={onChange} />)
    const input = wrapper.find('input')

    input.simulate('change')
    expect(onChange).toBeCalledTimes(1)
  })
})
