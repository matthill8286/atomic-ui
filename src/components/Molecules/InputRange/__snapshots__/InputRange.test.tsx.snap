// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputRange should render 1`] = `
.c4 {
  color: #00C8DC;
  font-family: "Oswald","Raleway",Arial,sans-serif;
  font-weight: 400;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c0 {
  padding: 24px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c1 {
  height: 6rem;
}

.c2 {
  width: 4rem;
  height: 4rem;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c3 {
  box-sizing: border-box;
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 2px solid #00C8DC;
  border-radius: 50%;
}

.c5 {
  font-size: 26px;
  -webkit-transform: translateY(0);
  -ms-transform: translateY(0);
  transform: translateY(0);
  line-height: 40px;
  padding: 0;
}

.c6 {
  -webkit-appearance: none;
  box-shadow: 0 1px 1px 0 rgba(0,0,0,0.07),0 2px 1px -1px rgba(0,0,0,0.06),0 1px 3px 0 rgba(0,0,0,0.10);
  height: 1px;
  width: 1px;
  background: #00C8DC;
  cursor: pointer;
  position: relative;
}

.c6:after {
  content: '';
  position: absolute;
  height: 24px;
  width: 24px;
  margin-left: -16px;
  border-radius: 50%;
  background: #FFFFFF;
  border: 5px solid #00C8DC;
}

.c7 {
  width: 100%;
  height: 32px;
  -webkit-appearance: none;
}

.c7:focus {
  outline: none;
}

.c7::-webkit-slider-thumb {
  -webkit-appearance: none;
  height: 1px;
  width: 1px;
  border-radius: 50%;
  border: 2px solid #FFFFFF;
  -webkit-transform: translateY(4px) scale(32);
  -ms-transform: translateY(4px) scale(32);
  transform: translateY(4px) scale(32);
  opacity: 0;
  cursor: pointer;
}

.c7::-moz-range-thumb {
  -webkit-appearance: none;
  height: 1px;
  width: 1px;
  border: 2px solid #FFFFFF;
  -webkit-transform: scale(32);
  -ms-transform: scale(32);
  transform: scale(32);
  border-radius: 50%;
  opacity: 0;
  border: none;
  cursor: pointer;
}

.c7::-ms-thumb {
  -webkit-appearance: none;
  height: 1px;
  width: 1px;
  border: 2px solid #FFFFFF;
  border-radius: 50%;
  opacity: 0;
  cursor: pointer;
}

.c7::-webkit-slider-runnable-track {
  width: 100%;
  height: 8px;
  cursor: pointer;
  background: #E5E5E5;
  box-shadow: 0 1px 1px 0 rgba(0,0,0,0.07),0 2px 1px -1px rgba(0,0,0,0.06),0 1px 3px 0 rgba(0,0,0,0.10);
  border-radius: 4px;
}

.c7:focus::-webkit-slider-runnable-track {
  background: #E5E5E5;
}

.c7::-moz-range-track {
  width: 100%;
  height: 8px;
  cursor: pointer;
  background: #E5E5E5;
  box-shadow: 0 1px 1px 0 rgba(0,0,0,0.07),0 2px 1px -1px rgba(0,0,0,0.06),0 1px 3px 0 rgba(0,0,0,0.10);
  border-radius: 4px;
}

.c7::-ms-track {
  width: 100%;
  height: 8pxpx;
  cursor: pointer;
  background: transparent;
  border-color: transparent;
  border-width: 16px 0;
  color: transparent;
}

.c7::-ms-fill-lower {
  background: #E5E5E5;
  box-shadow: 0 1px 1px 0 rgba(0,0,0,0.07),0 2px 1px -1px rgba(0,0,0,0.06),0 1px 3px 0 rgba(0,0,0,0.10);
  border-radius: 4px;
}

.c7::-ms-fill-upper {
  background: #E5E5E5;
  box-shadow: 0 1px 1px 0 rgba(0,0,0,0.07),0 2px 1px -1px rgba(0,0,0,0.06),0 1px 3px 0 rgba(0,0,0,0.10);
  border-radius: 4px;
}

.c7:focus::-ms-fill-lower {
  background: #E5E5E5;
}

.c7:focus::-ms-fill-upper {
  background: #E5E5E5;
}

@media (max-width:512px) {
  .c0 {
    padding: 24px 32px;
  }
}

@media (max-width:752px) {
  .c2 {
    -webkit-transform: translateX(-50%) translateY(16px);
    -ms-transform: translateX(-50%) translateY(16px);
    transform: translateX(-50%) translateY(16px);
  }
}

@media (max-width:752px) {
  .c5 {
    font-size: 18px;
  }
}

@media (max-width:752px) {
  .c7::-webkit-slider-thumb {
    -webkit-transform: translateY(4px) scale(16);
    -ms-transform: translateY(4px) scale(16);
    transform: translateY(4px) scale(16);
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <output
      className="c2"
      style={
        Object {
          "left": "100%",
        }
      }
    >
      <output
        className="c3"
      >
        <span
          className="c4 c5"
          color="primary"
          fontFamily="default"
        >
          10
        </span>
      </output>
    </output>
  </div>
  <div
    className="c6"
    style={
      Object {
        "left": "100%",
      }
    }
  />
  <input
    className="c7"
    list="input-range"
    max="1"
    min="0"
    name="input-range"
    onChange={[Function]}
    onMouseDown={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    step="0.1"
    type="range"
    value="1"
  />
</div>
`;
